add fees support for mainnet transaction

bisakhmondal opened the pull request:
A brief description of the changes in this PR:

closes: #85 

Incorporate
- `withFee` flag
- singleton `transactionFee` get method
- (TODO): in absence of withFee fallback to min_fee captured from blockchain

Tasks to complete before merging PR:
- [x]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/zwalletcli/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

Associated PRs (Link as appropriate):
- blobber:
- gosdk: https://github.com/0chain/gosdk/pull/623
- system_test:
- zboxcli:
- 0chain: https://github.com/0chain/0chain/pull/1753
- Other: ...

bisakhmondal committed:
add necessary logic

bisakhmondal committed:
update implementations with transaction fee

bisakhmondal committed:
revert unnecessary errors & gofmt

dabasov commented on cmd/root.go:
```diff
@@ -21,6 +21,11 @@ var cDir string
 
+// MinTxFee sets the min tx fee that must be paid for a tx to occur.
+// making it public for the flexibility to set it through ldflags.
+var MinTxFee float64
```
we should use int64 in SAS

dabasov commented on cmd/root.go:
```diff
@@ -21,6 +21,11 @@ var cDir string
+// MinTxFee sets the min tx fee that must be paid for a tx to occur.
+// making it public for the flexibility to set it through ldflags.
+var MinTxFee float64
+var txFee float64
```
the same int64 in SAS

dabasov reviewed (approved):
lgtm

dabasov reviewed (approved):
lgtm

dabasov commented:
please fix tests @bisakhmondal

bisakhmondal commented:
> please fix tests @bisakhmondal

On it.

bisakhmondal committed:
update with gosdk#623

bisakhmondal commented:
> please fix tests @bisakhmondal

Hey @dabasov, I think we have to deploy the 0chain core to devnet for the estimate_tx_cost API
![image](https://user-images.githubusercontent.com/41498427/198390898-bd133ac5-bd9f-4ff2-98ab-cbcdcc2c5776.png)

Any heads up on how to do that?

bisakhmondal committed:
replace ldflag to cliflag for minfee

dabasov reviewed (approved):
lgtm

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee-integration  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Update gosdk dep

peterlimg committed:
Fix go mod conflicts

peterlimg committed:
Refactor to simplify the changes

- Avoid doing token converting on each new transaction calling, convert it once on parsing the txn fee flag
- Update gosdk to do txn fee adjustment on gosdk end, so the user don't have to do it manually.
- Remove unsupported/unused velocity vars and methods

peterlimg committed:
Merge branch 'staging' into feat/chain-fee

peterlimg committed:
Update txn fee

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Update gosdk for txn fee

peterlimg committed:
Merge branch 'staging' into feat/chain-fee

* staging:
  upgrade GoSDK to v1.8.11
  fix: wrong gosdk version usage fixed
  changed stakepool lock/unlock interface (#199)
  fix: removed usage of local gosdk
  feature: added remove authorizer method

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Merge branch 'staging' into feat/chain-fee

* staging:
  upgrade GoSDK to v1.8.12
  chore(gomod): upgraded gosdk
  chore(gomod): upgraded gosdk
  chore(gomod): upgraded gosdk
  chore(gomod): upgraded gosdk
  chore(gomod): upgraded gosdk
  chore(gomod): upgraded gosdk
  chore(gomod): upgraded gosdk
  chore(gomod): upgraded gosdk
  chore(gomod): upgraded gosdk
  chore(gomod): upgraded gosdk
  chore(git): merge staging
  chore(git): merge staging
  fix(zcncore): used zcncore.MakeSCRestAPICall instead of zboxutil.MakeSCRestAPICall

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Fix init wallet read pool create

boddumanohar commented:
@dabasov are these changes good to merge?

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Merge branch 'staging' into feat/chain-fee

# By Yaroslav Svitlytskyi (10) and others
# Via GitHub (6) and Aniket Singh (1)
* staging:
  update gosdk to latest commit
  update gosdk to latest commit
  make changes corresponding to changes in balance package in gosdk
  Update tests.yml
  Update system_tests.yml
  Update system_tests.yml
  Update tests.yml
  Update system_tests.yml
  Update tests.yml
  Update system_tests.yml
  Update system_tests.yml
  Update tests.yml
  Update system_tests.yml
  feature: updated system tests workflow
  update gosdk to staging
  updated gosdk version
  Revert "updated gosdk version"
  updated gosdk version
  feat: add create-wallet cli command
  refactor: code simplification

# Conflicts:
#	cmd/getbalance.go
#	cmd/root.go
#	cmd/send.go
#	go.mod
#	go.sum

peterlimg committed:
Update gosdk

Kishan-Dhakan committed:
Merge branch 'staging' into feat/chain-fee

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Merge branch 'staging' into feat/chain-fee

peterlimg committed:
Update gosdk

peterlimg committed:
Update gosdk

peterlimg committed:
Update gosdk

peterlimg committed:
Update gosdk to fix zero balance

Kishan-Dhakan committed:
resolve conflicts, update gosdk

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | ABILITY TO SKIP TESTS IS DISABLED. SKIP_TESTS OPTION IGNORED.  |
| system-tests  | chain-fee  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | ABILITY TO SKIP TESTS IS DISABLED. SKIP_TESTS OPTION IGNORED.  |
| system-tests  | chain-fee-remove-excess-faucet  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | ABILITY TO SKIP TESTS IS DISABLED. SKIP_TESTS OPTION IGNORED.  |
| system-tests  | chain-fee  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Update gosdk

peterlimg committed:
Merge branch 'staging' into feat/chain-fee

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Merge branch 'staging' into feat/chain-fee

* staging:
  Fix: reconciliation problem (#222)

peterlimg committed:
Merge branch 'staging' into feat/chain-fee

* staging:
  Feature: modify "GetAuthorizers" function for the latest interface (#228)

peterlimg committed:
Update gosdk

peterlimg committed:
Update gosdk

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Update gosdk dep

peterlimg committed:
Update gosdk

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Update gosdk

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee-v2  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0dns  | staging  |

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | chain-fee  |
| gosdk  | chain-fee  |
| zboxcli  | chain-fee  |
| zwalletcli  | feat/chain-fee  |
| 0chain  | chain-fee-v2  |
| blobber  | chain-fee  |
| authorizer  | chain-fee  |
| 0box  | chain-fee  |
| 0dns  | staging  |
